services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PinnedFlowSpec.groovy

```diff
@@ -0,0 +1,111 @@
+            [sw.dpId, northbound.getSwitchRules(sw.dpId).flowEntries.findAll {
+                !Cookie.isDefaultRule(it.cookie)
+            }*.cookie]
+        }
+        def metersMap = involvedSwitches.collectEntries { sw ->
+            [sw.dpId, northbound.getAllMeters(sw.dpId).meterEntries.findAll {
+                it.meterId > MAX_SYSTEM_RULE_METER_ID
+            }*.meterId]
+        }
+
+        northbound.portDown(islToBreak.srcSwitch.dpId, islToBreak.srcPort)
+
+        then: "Flow is not rerouted and marked as DOWN"
+        Wrappers.wait(WAIT_OFFSET) {
+            assert northbound.getFlowStatus(flow.id).status == FlowState.DOWN
+            assert pathHelper.convert(northbound.getFlowPath(flow.id)) == currentPath
+        }
+
+        and: "Rules and meters are not changed"
+        def cookiesMapAfterReroute = involvedSwitches.collectEntries { sw ->
+            [sw.dpId, northbound.getSwitchRules(sw.dpId).flowEntries.findAll {
+                !Cookie.isDefaultRule(it.cookie)
+            }*.cookie]
+        }
+        def metersMapAfterReroute = involvedSwitches.collectEntries { sw ->
+            [sw.dpId, northbound.getAllMeters(sw.dpId).meterEntries.findAll {
+                it.meterId > MAX_SYSTEM_RULE_METER_ID
+            }*.meterId]
+        }
+
+        cookiesMap.sort() == cookiesMapAfterReroute.sort()
+        metersMap.sort() == metersMapAfterReroute.sort()
+
+        when: "The broken ISL is restored"
+        northbound.portUp(islToBreak.srcSwitch.dpId, islToBreak.srcPort)
+
+        then: "The flow is marked as UP and not rerouted"
+        Wrappers.wait(WAIT_OFFSET + discoveryInterval) {
+            assert islUtils.getIslInfo(islToBreak).get().state == IslChangeType.DISCOVERED
+            assert northbound.getFlowStatus(flow.id).status == FlowState.UP
+            assert pathHelper.convert(northbound.getFlowPath(flow.id)) == currentPath
+        }
+
+        and: "Cleanup: revert system to original state"
+        flowHelper.deleteFlow(flow.id)
+        database.resetCosts()
+    }
+
+    def "System is able to reroute(intentional) pinned flow"() {
+        given: "A pinned flow with alt path available"
+        def switchPair = topologyHelper.getAllNeighboringSwitchPairs().find { it.paths.size() > 1 } ?:
+                assumeTrue("No suiting switches found", false)
+        def flow = flowHelper.randomFlow(switchPair)
+        flow.pinned = true
+        flowHelper.addFlow(flow)
+
+        def currentPath = pathHelper.convert(northbound.getFlowPath(flow.id))
+
+        when: "Make another path more preferable"
+        def newPath = switchPair.paths.find { it != currentPath }
+        switchPair.paths.findAll { it != newPath }.each { pathHelper.makePathMorePreferable(newPath, it) }
+
+        and: "Init reroute(manually)"
+        def isl = pathHelper.getInvolvedIsls(currentPath).first()
+        northbound.rerouteLinkFlows(isl.srcSwitch.dpId, isl.srcPort, isl.dstSwitch.dpId, isl.dstPort)
```

Comment:
why not `northbound.rerouteFlow(flow.id)` ? This is fine, but just wondering

Reply:
no specific reason.
just tried to do it in unusual way